### PR TITLE
DAML-LF internal type safety

### LINKED ISSUE #398: Submitter supplied information in commands is unrestricted UTF-8
The ledger-api command contains the fields workflow_id, application_id and command_id and places no restrictions on the contents of these strings. Since with the DAML-on-X approach there are going to be multiple different backend implementations that will require correctly handling these 
(serializing, storing in database), it would be safer to limit the range of these fields to an ASCII subset (for example as defined by SimpleString).

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR continue the effort of #705  by cleaning up type in the Scala code
It focus on string used as Id in the ledger

- ContractId
- LedgerId
- TransactionId
- CommandId
- WorkflowId
- ApplicationId
- EventId

This also fixes  #398

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by remyhaemmerle-da** [comment]
@gaboraranyossy-da Could you take a look  at `PostgresLedgerDao` and co ?

**Comment by remyhaemmerle-da** [comment]
@bitonic I have done the changes you requested

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala [hunk 1]
```diff
@@ -1067,6 +1067,9 @@ object EngineTest {
   private implicit def toParty(s: String): Party =
     Party.assertFromString(s)
 
+  private implicit def toContractId(s: String): ContractIdString =
+    LedgerString.assertFromString(s)
```
Comment (S11001001) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala]:
```suggestion
    ContractIdString.assertFromString(s)
```

As a consequence of the proposed aliases for `LedgerString`, and similarly for other direct references to `LedgerString` in order to call methods.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala [hunk 2]
```diff
@@ -39,6 +42,77 @@ object MatchingStringModule extends (Regex => MatchingStringModule) {
...
+  final def fromInt(i: Int): T = fromLong(i.toLong)
+
+  final def concat(s: T, ss: T*): T = {
+    val b = newBuilder
+    b += s
+    ss.foreach(b += _)
+    b.result()
+  }
+
+  def newBuilder: mutable.Builder[T, T]
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala]: indeed. Good catch

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala [hunk 3]
```diff
@@ -39,6 +42,77 @@ object MatchingStringModule extends (Regex => MatchingStringModule) {
...
+
+      def fromLong(i: Long): T = i.toString
+
+      def equalInstance: Equal[T] = scalaz.std.string.stringInstance
+
+      val Array: ArrayFactory[T] = new ArrayFactory[T]
+
+      def toStringMap[V](map: Map[T, V]): Map[String, V] = map
+
+      def newBuilder: mutable.ReusableBuilder[T, T] = new mutable.ReusableBuilder[T, T] {
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala]: done

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala [hunk 4]
```diff
@@ -39,6 +42,77 @@ object MatchingStringModule extends (Regex => MatchingStringModule) {
...
+            .fold[Either[String, T]](Right(s))(c =>
+              Left(s"""non expected character 0x${c.toInt.toHexString} in "$s""""))
+
+      def fromLong(i: Long): T = i.toString
+
+      def equalInstance: Equal[T] = scalaz.std.string.stringInstance
+
+      val Array: ArrayFactory[T] = new ArrayFactory[T]
+
+      def toStringMap[V](map: Map[T, V]): Map[String, V] = map
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala]: done

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala [hunk 5]
```diff
@@ -157,4 +157,21 @@ object Ref {
...
+  /**
+    * Used to reference to leger objects like contractIds, ledgerIds,
+    * transactionId, ... We use the same type for those ids, because we
+    * construct some by concatenating the others.
+    */
+  val LedgerString = ConcatenableMatchingStringModule("._:-#".contains(_), 256)
+  type LedgerString = LedgerString.T
+
+  /** Identifier for a contractId */
+  type ContractIdString = LedgerString
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala]: done

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala]: mostly done. I have some more cleanup to do. I will fix the remaining ones (EventId, WorkflowId, ApplicationId ...)  in an upcoming PR.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/util/Conversions.scala [hunk 6]
```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (gaboraranyossy-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/util/Conversions.scala]: they look reasonable to me. Most likely has no impact, but could you compare a perf-test result vs master locally?

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala [hunk 7]
```diff
@@ -39,6 +42,77 @@ object MatchingStringModule extends (Regex => MatchingStringModule) {
...
+object ConcatenableMatchingStringModule {
+
+  def apply(
+      extraAllowedChars: Char => Boolean,
+      maxLength: Int = Int.MaxValue
+  ): ConcatenableMatchingStringModule =
+    new ConcatenableMatchingStringModule {
+      type T = String
+
+      def fromString(s: String): Either[String, T] =
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala]: OK

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala [hunk 8]
```diff
@@ -24,21 +22,84 @@ sealed abstract class MatchingStringModule {
...
+            .fold[Either[String, T]](Right(s))(c =>
+              Left(s"""non expected character 0x${c.toInt.toHexString} in "$s""""))
+
+      override def fromLong(i: Long): T = i.toString
+
+      override def equalInstance: Equal[T] = scalaz.std.string.stringInstance
+
+      override val Array: ArrayFactory[T] = new ArrayFactory[T]
+
+      override def toStringMap[V](map: Map[T, V]): Map[String, V] = map
```
Comment (S11001001) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala]: Similarly to the other `toStringMap` here.

Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala]: done

**Pull request merged**